src/element/linearElementEditor.ts

```diff
@@ -0,0 +1,360 @@
+        // +1px to account for outline stroke
+        this.POINT_HANDLE_SIZE / 2 + 1
+      ) {
+        return idx;
+      }
+    }
+    return -1;
+  }
+
+  static createPointAt(
+    element: NonDeleted<ExcalidrawLinearElement>,
+    scenePointerX: number,
+    scenePointerY: number,
+  ): Point {
+    const [x1, y1, x2, y2] = getElementAbsoluteCoords(element);
+    const cx = (x1 + x2) / 2;
+    const cy = (y1 + y2) / 2;
+    const [rotatedX, rotatedY] = rotate(
+      scenePointerX,
+      scenePointerY,
+      cx,
+      cy,
+      -element.angle,
+    );
+
+    return [rotatedX - element.x, rotatedY - element.y];
+  }
+
+  // element-mutating methods
+  // ---------------------------------------------------------------------------
+
+  /**
+   * Normalizes line points so that the start point is at [0,0]. This is
+   *  expected in various parts of the codebase.
+   */
+  static normalizePoints(element: NonDeleted<ExcalidrawLinearElement>) {
+    const { points } = element;
+
+    const offsetX = points[0][0];
+    const offsetY = points[0][1];
+
+    mutateElement(element, {
+      points: points.map((point, _idx) => {
+        return [point[0] - offsetX, point[1] - offsetY] as const;
+      }),
+      x: element.x + offsetX,
+      y: element.y + offsetY,
+    });
+  }
+
+  static movePoint(
+    element: NonDeleted<ExcalidrawLinearElement>,
+    pointIndex: number,
+    targetPosition: Point,
+  ) {
+    const { points } = element;
+
+    // in case we're moving start point, instead of modifying its position
+    //  which would break the invariant of it being at [0,0], we move
+    //  all the other points in the opposite direction by delta to
+    //  offset it. We do the same with actual element.x/y position, so
+    //  this hacks are completely transparent to the user.
+    let offsetX = 0;
+    let offsetY = 0;
+
+    const deltaX = targetPosition[0] - points[pointIndex][0];
+    const deltaY = targetPosition[1] - points[pointIndex][1];
+    const nextPoints = points.map((point, idx) => {
+      if (idx === pointIndex) {
+        if (idx === 0) {
+          offsetX = deltaX;
+          offsetY = deltaY;
+          return point;
+        }
+        offsetX = 0;
+        offsetY = 0;
+
+        return [point[0] + deltaX, point[1] + deltaY] as const;
+      }
+      return offsetX || offsetY
+        ? ([point[0] - offsetX, point[1] - offsetY] as const)
+        : point;
+    });
+
+    const nextCoords = getElementPointsCoords(element, nextPoints);
+    const prevCoords = getElementPointsCoords(element, points);
+    const nextCenterX = (nextCoords[0] + nextCoords[2]) / 2;
+    const nextCenterY = (nextCoords[1] + nextCoords[3]) / 2;
+    const prevCenterX = (prevCoords[0] + prevCoords[2]) / 2;
+    const prevCenterY = (prevCoords[1] + prevCoords[3]) / 2;
+    const dX = prevCenterX - nextCenterX;
+    const dY = prevCenterY - nextCenterY;
+    const rotated = rotate(offsetX, offsetY, dX, dY, element.angle);
```

Comment:
Can't explain what this means. I did simplify a complex equation which became like this.